src/NuGet.Jobs.Catalog2Registration/Hives/HiveStorage.cs

```diff
@@ -0,0 +1,434 @@
+                    path,
+                    blob.Properties.ContentEncoding);
+
+                return result;
+            }
+            catch (StorageException ex) when (ex.RequestInformation?.HttpStatusCode == (int)HttpStatusCode.NotFound)
+            {
+                _logger.LogInformation(
+                    "No blob in container {Container} at path {Path} exists.",
+                    GetContainerName(hive),
+                    path,
+                    blob.Properties.ContentEncoding);
+
+                if (allow404)
+                {
+                    return default(T);
+                }
+                else
+                {
+                    throw;
+                }
+            }
+            finally
+            {
+                _throttle.Release();
+            }
+        }
+
+        private MemoryStream Serialize<T>(HiveType hive, T entity)
+        {
+            var serializedStream = new MemoryStream();
+            Stream writeStream;
+            GZipStream gzipStream;
+            if (IsGzipped(hive))
+            {
+                gzipStream = new GZipStream(
+                    serializedStream,
+                    CompressionMode.Compress,
+                    leaveOpen: true);
+                writeStream = gzipStream;
+            }
+            else
+            {
+                gzipStream = null;
+                writeStream = serializedStream;
+            }
+
+            using (var streamWriter = new StreamWriter(
+                writeStream,
+                new UTF8Encoding(encoderShouldEmitUTF8Identifier: false),
+                bufferSize: 1024,
+                leaveOpen: true))
+            using (var jsonTextWriter = new JsonTextWriter(streamWriter))
+            {
+                NuGetJsonSerialization.Serializer.Serialize(jsonTextWriter, entity);
+            }
+
+            gzipStream?.Dispose();
+            serializedStream.Position = 0;
+            return serializedStream;
+        }
+
+        private async Task WriteAsync<T>(
+            HiveType hive,
+            IReadOnlyList<HiveType> replicaHives,
+            string path,
+            T entity,
+            string typeName,
+            Action<T, HiveType, HiveType> updateUrls)
+        {
+            var hiveToWriteTask = new Dictionary<HiveType, Task>();
+            var previousHive = hive;
+            foreach (var currentHive in GetHiveSequence(hive, replicaHives))
+            {
+                if (hiveToWriteTask.ContainsKey(currentHive))
+                {
+                    continue;
+                }
+
+                if (currentHive != previousHive)
+                {
+                    // The hive defines the base URL. If we're now writing to another hive (which is the case if this
+                    // is the 2nd ... Nth hive) then we need to update the URLs to point to the current hive.
+                    updateUrls(entity, previousHive, currentHive);
+                }
```

Comment:
This code assumes that each invocation of `updateUrls` will replace all properties on the entity. While this is true today, I'm concerned that this may break if we ever set a property only on a certain hive that is a replica of another hive.

For example, let's say I have hive `A` and replica hives `B` and `C`. I change `UpdateXUrls` builder so I update the property `newUrl` **only** when `toHive` is `B`. This would have the unexpected side effect of setting `newUrl` for hive `C` (and the entity won't be returned to the original state once `WriteAsync` completes). This seems like a subtle bug that one may easily miss.

Do you think this concern is worth addressing? One possible solution may be to clone the entity before updating its URLs.

Reply:
I considering a "clone" approach too but a properly written clone method needs to copy all properties of the POCO over as well. I thought it would be easier to get the "update URLs" method right than a "copy all properties" method since the former is less likely to be affected by future changes (since not all model changes are URL additions).

I think some other alternatives are:

1. Abstract URL setting so that both the initial population of the properties and the converting flow use the same data source (i.e. one method called in two different contexts)
1. Change the URL setting flow completely so one module has a "set URLs" method that is invoked for `A`, `B`, and `C` here (i.e one method called in one, low level context)

I'll think about this some more. I thought this would be enough but if you're concerned about it, it's probably worth rethinking.

Reply:
> I thought it would be easier to get the "update URLs" method right than a "copy all properties" method since the former is less likely to be affected by future changes (since not all model changes are URL additions).

That's a good point. Another suggestion: could you test all "from" and "to" hive combinations on all `updateUrls` handlers and ensure that all properties of the POCO are back to the original state?

Reply:
Haha, that's exactly the path I was going. And I recurse on the JSON object model finding all properties that look like URLs and assert the expected set.